subsys/bluetooth/services/ras/rreq/ras_rreq.c

```diff
@@ -0,0 +1,714 @@
+	if (err) {
+		LOG_DBG("On-demand ranging data subscribe failed (err %d)", err);
+		return err;
+	}
+
+	LOG_DBG("On-demand ranging data subscribed");
+
+	return 0;
+}
+
+static int ranging_data_ready_subscribe(struct bt_ras_rreq *rreq, bt_ras_rreq_rd_ready_cb_t cb)
+{
+	int err;
+
+	err = bt_gatt_subscribe(rreq->conn, &rreq->rd_ready.subscribe_params);
+	if (err) {
+		LOG_DBG("Ranging data ready subscribe failed (err %d)", err);
+		return err;
+	}
+
+	LOG_DBG("Ranging data ready subscribed");
+	rreq->rd_ready.cb = cb;
+
+	return 0;
+}
+
+static int ranging_data_overwritten_subscribe(struct bt_ras_rreq *rreq,
+					      bt_ras_rreq_rd_overwritten_cb_t cb)
+{
+	int err;
+
+	err = bt_gatt_subscribe(rreq->conn, &rreq->rd_overwritten.subscribe_params);
+	if (err) {
+		LOG_DBG("Ranging data overwritten subscribe failed (err %d)", err);
+		return err;
+	}
+
+	LOG_DBG("Ranging data overwritten subscribed");
+	rreq->rd_overwritten.cb = cb;
+
+	return 0;
+}
+
+void bt_ras_rreq_free(struct bt_conn *conn)
+{
+	int err = 0;
+	struct bt_conn_info info;
+	struct bt_ras_rreq *rreq = ras_rreq_find(conn);
+
+	if (!rreq) {
+		return;
+	}
+
+	LOG_DBG("Free rreq %p for conn %p", (void *)conn, (void *)rreq);
+
+	err = bt_conn_get_info(conn, &info);
+	if (err != 0) {
+		bt_conn_unref(rreq->conn);
+		rreq->conn = NULL;
+		return;
+	}
+
+	if (info.state == BT_CONN_STATE_CONNECTED) {
+		/* -EINVAL means already not subscribed */
+		err = bt_gatt_unsubscribe(conn, &rreq->cp.subscribe_params);
+		if (err != 0 && err != -EINVAL) {
+			LOG_DBG("Failed to unsubscribe to RAS-CP: %d", err);
+		}
+
+		err = bt_gatt_unsubscribe(conn, &rreq->on_demand_rd.subscribe_params);
+		if (err != 0 && err != -EINVAL) {
+			LOG_DBG("Failed to unsubscribe to ondemand ranging data: %d", err);
+		}
+
+		err = bt_gatt_unsubscribe(conn, &rreq->rd_ready.subscribe_params);
+		if (err != 0 && err != -EINVAL) {
+			LOG_DBG("Failed to unsubscribe to ranging data ready: %d", err);
+		}
+
+		err = bt_gatt_unsubscribe(conn, &rreq->rd_overwritten.subscribe_params);
+		if (err != 0 && err != -EINVAL) {
+			LOG_DBG("Failed to unsubscribe to ranging data overwritten: %d", err);
+		}
+	}
+
+	bt_conn_unref(rreq->conn);
+	rreq->conn = NULL;
+}
+
+int bt_ras_rreq_on_demand_ranging_data_subscribe_all(
+	struct bt_conn *conn, bt_ras_rreq_rd_ready_cb_t rd_ready_cb,
+	bt_ras_rreq_rd_overwritten_cb_t rd_overwritten_cb)
+{
+	int err;
+	struct bt_ras_rreq *rreq = ras_rreq_find(conn);
+
+	if (rreq == NULL) {
+		return -EINVAL;
+	}
+
+	err = ras_rreq_cp_subscribe(rreq);
+	if (err) {
+		return err;
+	}
+
+	err = on_demand_ranging_data_subscribe(rreq);
+	if (err) {
+		return err;
+	}
+
+	err = ranging_data_ready_subscribe(rreq, rd_ready_cb);
+	if (err) {
+		return err;
+	}
+
+	err = ranging_data_overwritten_subscribe(rreq, rd_overwritten_cb);
+	if (err) {
+		return err;
+	}
+
+	return 0;
+}
+
+int bt_ras_rreq_alloc_and_assign_handles(struct bt_gatt_dm *dm, struct bt_conn *conn)
```

Comment:
Similar question here. If this fails after allocating the `rreq`, do we expect the app to free the `rreq` or should we do it ourselves?

Reply:
Yeah maybe we should free the rreq